### DISCUSSION TITLE
Performance

### DISCUSSION THREAD

**Discussion by EdwHopper** [blackears/blenderNormalBrush#4] [General]
Hi, I am testing this great idea with some scan meshes that I need to adjust. The operation lags at almost .5fps even when just selecting a UI option, everywhere in blender, let alone brush strokes. The meshes I use have 100-200k tris with custom normals and I use blender 3.0 stable and the lastest release of the addon. I wander if there is something I can do to improve performance. I already tried to disable overlays but it has no effect. Thanks!

**Comment by blackears** [comment]
You could try splitting your mesh into chunks and using the normals brush on each chunk before putting them pack together.  Unfortunately the normals brush plugin runs in Python, so I can't make it nearly as efficient as something that would run natively.

**Reply by EdwHopper** [reply]
Thanks for answer! I know python is slow but is it not possible to optimize it somehow? I feel something is wrong with the fact all blender interface freezes even while I don't hover the viewport. I believe there is a good reason it's made to work in object mode but maybe edit mode offers other advantages like updating just the verts under the cursor? Or maybe a delayed execution option, until a stroke is finished? I'm just tring to understand how it works and maybe I can suggest something useful.

**Comment by blackears** [comment]
I optimized it as much as I could when I wrote it.  Maybe geometry nodes would offer another way to do things, but I'm still learning how to use those.  It's possible the model you're trying to use is different from the ones I used for testing - if you could send me something to examine, I could make a better judgement about what's going wrong.  A checkbox for delayed execution on the stroke might be a good idea.

Blender has two very different systems for editing meshes in object mode and edit mode.  However, both involve committing the entire object after you've made your changes.  I don't think rewriting it to use edit mode would change the update speed.

**Comment by blackears** [comment]
Also, regarding the interface freezing when the tool is in use - this is a problem all tool plugins have.  Whenever a tool is active, the UI panel is unable to receive any input.  I've posted about this several times on forums (and other plugin developers too), but the Blender staff doesn't seem interested in fixing this.  The only plugin I'm aware of that doesn't have this problem is Retopoflow, and that's because they wrote their own UI system from the ground up.

The best advice I can give you is to make sure to press Enter to commit your changes before trying to use the panel.

**Comment by blackears** [comment]
I've made some changes to the addon which improve the rendering speed, but I can't so very much about how long it takes the mesh to update.  Try the 1.1.9 version and let me know if it's an improvement.

**Comment by EdwHopper** [comment]
Ok just to get clear about the freezing UI issue with highpoly meshes, nothing is completeley stuck but it takes like 10 seconds or more to register a change. And it's not only on the addon's panel but everywhere in blender's ui.
I will note here the rest of the observations I have, testing on the monkey mesh.
- while the addon is running(paint on), the mode buttons are not registering the clicks. Click once behaves like click+drag over the buttons and you need a second click to stop it. It happens also over blenders own Properties panel's icons. This behavior is confusing.
- if part of the object is under the addon's popup, while painting, the addon ui is completely ignoring the input. This might be related to above. 
- working with centimeter scale objects (10cm subdivided cube), scene units cm and unit scale 1, the brushes work inconsistently. This happens with other tools in blender so it's not a big problem.
- lastly, I tried your suggestion to split separate small patches of the highpoly meshes and it works. But if I try Copy seams normals is just taking forever and I need to force restart blender to stop it. Maybe a warning about slow performance with dense meshes would help users.

**Comment by blackears** [comment]
You first two points are things I have no control over.  The UI system Blender provides for controls was designed for things that just run once when you press a button, not for tools that keep working over a period of time.  The only way around this is to design your own completely independent UI system.

Not sure why copy seam is doing that.  Might need to have  another look at it.  In the mean time you can try just joining the meshes normally and then brushing over the seams to fix the normals that way.
